README
------------------------------------------
deprec_bunsi - updated version from deprec
------------------------------------------

== Info

This is an updated version of Mike Bailey's deprec. It basically adds a few helpful new deployment recipes and updates some functions to better serve installation and deployment of bunsi.net, which, however, are most likely as helpful to other websites as they are to bunsi.net

== Install

1. Check out from Github
2. build gem from tree: 
   gem build deprec.gemspec 
3. install newly build gem from local folder:
   sudo gem install --no-rdoc --no-ri -f -l deprec_bunsi-<current_version>.gem


------------------------------------------
deprec - Deployment Recipes for Capistrano
------------------------------------------

== Introduction

The deprec [1] gem is a set of tasks for Capistrano [2]. These tasks provide
for the installation, configuration and control of system services on servers
running Ubuntu linux. Deprec was created in 2006 by Mike Bailey to setup an
environment for running Ruby on Rails web applications on Ubuntu servers. Since
then its uses have grown to installing Xen virtualization, mail, monitoring, high
availability IP failover and other services.

The tasks are run at the command line on your workstation and connect to 
remote servers via ssh to run commands and copy out files.

Deprec-2.x is a complete rewrite of the project that achieves the following:

- support for Capistrano 2
- generated config files are stored locally to enable editing and version control
- support for more services (heartbeat, nagios, nginx, ntp, postfix, etc) 
- multiple Rails deployment options (Passenger+Apache, Mongrel+Apache/Nginx)
- creation of a standard base set of task names
- tasks are cleanly separated into namespaced units (one file per service)
- interactive prompting for missing config values

Deprec and Capistrano are written in the Ruby programming language [3] however 
no knowledge of Ruby is required to use it. Users should be able to write 
new tasks and modify existing options without prior knowledge of Ruby.


== Installation

Deprec can be obtained from rubyforge[4] and installed using rubygems[5].

	sudo gem install deprec  # installs deprec and dependancies 
	cap depify .			 # creates ~/.caprc which you may edit
	cap -T					 # should list lots of deprec tasks

The .caprc file is loaded every time you use Capistrano. It in turn loads 
the deprec tasks so you always have them available. Editing the .caprc file 
in your home directory allows you to specify the location of your ssh key
and enable some other useful options (documented in the comments). You can
also put tasks here that you want access to, regardless of the current working
directory.


== Getting a Ruby on Rails app running on a fresh Ubuntu server

This is still what brings people to deprec. You can install a full Rails stack
and get multiple apps running on it in much less time than it would take to 
do it manually (e.g. an hour vs. a weekend).

Get your app running on a fresh Ubuntu (7.10, 8.04) server with these commands:

    cd your_rails_app
    depify . 
    # Edit config/deploy.rb
    cap deprec:rails:install_stack

	# WARNING! Don't run the following command if you are using a shared 
	# database server that has already been installed.
	cap deprec:db:install 

    cap deploy:setup 
    cap deploy 
    cap deploy:migrate

Recorded times for complete install on Ubuntu hardy server (7.10 amd64)

	* Passenger+REE+Apache        : 14m29s
	* Passenger+ruby-1.8.7+Apache : 17m51s
	* nginx+mongrel               : 16m25s
	
Note you can choose combinations of:
	
	* Ruby Enterprise Edition or ruby-1.8.7
	* Passenger or mongrel (passenger requires you choose to the apache option)
	* Apache or nginx

You can find documentation on the deprec site. http://www.deprec.org/


== Installing other things

I plan to document other things I use deprec for on http://www.deprec.org/. 
Feel free to poke around and see what's there. I use deprec to provision and 
manage servers so you might find some things in there I haven't documented. Lucky you.


== Disclaimer

The tasks run commands that may make changes to your workstation and remote server. 
You are advised to read the source and use at your own risk.


== Credits

Deprec is written and maintained by Mike Bailey <mike@bailey.net.au>. 
More about me here: [http://mike.bailey.net.au/]

Deprec was inspired and uses the brilliantly executed Capistrano. Thanks Jamis!
This gem includes a modified copy of Neil Wilson's very useful vmbuilder_plugins gem.


== Thanks

Deprec wouldn't be what it is without the contributions of many people, a few of whom are listed here:

  Square Circle Triangle: commissioned work that has included in the project.
  Eric Harris-Braun: great testing, bug reports and suggestions
  Gus Gollings: helped restore www.deprec.org
  Craig Ambrose: testing, documentation and beer

  github forks of note:
    isaac
    paulreimer
    jasherai
    saimonmoore
    zippy


== License

Deprec is licenced under the GPL. This means that you can use it in commercial 
or open source applications. More details found here:
http://www.gnu.org/licenses/gpl.html

deprec - deployment recipes for capistrano
Copyright (C) 2006-2008 Mike Bailey

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


[1] http://www.deprec.org/
[2] http://www.capify.org/
[3] http://www.ruby-lang.org/en/
[4] http://rubyforge.org/
[5] http://rubygems.org/
[6] http://www.sct.com.au/

foo


CHANGELOG
# dprec_bunsi changelog

= 2.2.0 (Jan 26, 2010)

* Added `mod_rails` recipe, which uses Nginx as web server and Passenger as app server 
* Added `phpmyadmin` recipe; uses version 3.2.4 
* Added `minicgi` recipe for PHP support on Nginx
* Updated REE to 1.8.7-2009.10 + addl. minor task changes in the recipe
* Updated Sphinx to 0.9.9 
* Updated Monit to 5.0.3
* Changed run method for MySQL; add task :change_root_pw

# deprec changelog

= 2.1.4 (Sep 16, 2009)

* rubygems-1.3.5 
* git-1.6.4.3 

= 2.1.2 (Aug 26, 2009)

* Better support for Capistrano multistage
  Look for database.yml in config/#{stage}/database.yml

= 2.1.1 (Aug 25, 2009)

* Added haproxy

= 2.1.0 (Aug 24, 2009)

* Added Collectd
* Added Syslog
* Updated Nagios to 3.2.0

== 2.0.15 (August 10, 2009)

* Moved apache logs for vhosts into /var/log/apache2 (override by setting :apache_log_dir)

== 2.0.14 (August 10, 2009)

* Added logrotate config for passenger

== 2.0.13 (July 5, 2009)

* Bugfix: updated url for REE (old one went away)

== 2.0.12 (June 24, 2009)

* Updated chef recipes

== 2.0.11 (June 19, 2009)

* Security update: ruby-1.8.7-p174

== 2.0.10 (June 14, 2009)

* Added CouchDB
* Added Chef
* Fixed Passenger gem version problem (permanently!) [gfairbrother]
* Updated ree version to ruby-enterprise-1.8.6-20090610

== 2.0.9 (April 14, 2009)

* Use passenger-2.1.3

* Added support for Capistrano multistage. Generated config files will be 
  have a directory with the stage name prefixed to their local path.

  e.g. mbailey$ cap production deprec:monit:config_gen

  [done] config/production/monit/etc/init.d/monit written
  [done] config/production/monit/etc/monitrc written
  [done] config/production/monit/etc/monit.d/nothing written

== 2.0.8 (March 17, 2009)

* :config task now renders configs directly to remote server if locally 
  rendered versions (from :config_gen task) do not exist.

== 2.0.7 (March 17, 2009)

* bugfix

== 2.0.6 (March 17, 2009)

* Added option to create ~/.caprc file.  run 'depify -c'

== 2.0.5 (March 17, 2009)

* Added recipes and config templates for Integrity ci. http://integrityapp.com/
* bugfix: need to enable apache mods on install

== 2.0.4 (March 16, 2009)

* bugfix - updated Passenger version number

== 2.0.3 (March 3, 2009)

* bugfix - can't set ownership of environment.rb before deploy!

== 2.0.2 (March 3, 2009)

* Change owner of config/environment.rb to the value of app_user after deploy
  so Passenger will run app as this user [mbailey]

* bugfix: set homedir for app_user account to value of deploy_to. 
  Passenger complains if homedir for user it runs as doesn't exist. [mbailey]

* Run the database rake tasks on the app server, not the db server. [mbailey]
  e.g. cap deprec:database:schema_load

* Updated to ruby-enterprise-1.8.6-20090201

* Scan for ssh public key if not provided locally


== 2.0.1 (February 20, 2009)

Newly created ~/.caprc does not load Rails recipes.

This solves the problem of Capistrano's default deploy tasks
having 'before' and 'after' tasks added when you're deploying
a project that isn't setup to use deprec.

Existing users can update their existing ~/.caprc, replacing:

    require 'deprec'

with:

    require 'deprec_minus_rails'

== 2.0.0 (February 14, 2009)

* Official deprec-2.0 Rubyforge release!
 
== 1.99.34 (February  6, 2009)

* Bugfix: restricted some tasks to certain roles

== 1.99.33 (February  5, 2009)

# Added options for :web, :app, :db and :ruby software to install/use.
#
# Newly generated deploy.rb files specify Passenger, Apache and Ruby Enterprise 
# Edition. If not specified, deprec defaults to Mongrel, Nginx and Ruby-1.8.7.
# This is so that existing users of deprec2 previews will not be affected.

* Added :ruby_vm_type option to choose between :mri and :ree Ruby versions

* Added :app_server_type
  :passenger [set in new deploy.rb]
  :mongrel   [default if not set]

* Added :web_server_type 
  :apache [set in new depoy.rb] (for :mongrel or :passenger apps)
  :nginx  [default if not set]  (for :mongrel apps) 

* Added :db_server_type
  :mysql [default]
  :postgresql
  :sqlite

# Other additions

* cap deprec:rails:setup_database

  Creates database and user for Rails app using data from database.yml.
  Respects value of :db_server_type

* Postgresql recipes [Eric Harris-Braun]

* Starling Recipes [Rob Caporetto]

* cap deprec:utils:handy 

    Installs a bunch of handy utils I can't live without:

    lynx nmap netcat telnet dnsutils rsync curl wget mailx mutt vim-full tree

* Updated Apache vhost template for mongrel. Supports https and puts in dummy
  certificate. Useful settings are:

  set :apache_ssl_enabled, true
  set :apache_ssl_forward_all, true

== 1.99.30 (November 28, 2008)

* Bugfix: use current_path for passenger_root

== 1.99.29 (November 28, 2008)

* Bugfix: deprec:passenger:install_deps was commented out for testing

== 1.99.28 (November 26, 2008)

* Added support for Phusion Passenger (http://www.modrails.com/) [mbailey]

* Restricted tasks to roles. [mbailey]
  http://github.com/mbailey/deprec/commit/d103dda7ee3f7696007d230c76c7739dd57115ec

* Extend capistrano_extensions.rb to allow install from git sources. [jasherai]

* Updated git to 1.6.0.1 [thornysun]
 
== 1.99.27 (August 27, 2008)

* Bugfix: Add configure and make options to src_package for rubygems to 
  override defaults. [jasherai]

* Change: We now setup postfix as a satellite mail server. This is designed 
  to listen only on the local network interface and non-local mail to a relay
  host. [mbailey]

== 1.99.26 (August 22, 2008)

* added vnstat recipes - measure and report traffic on interface counters [mbailey]
* upgrade ruby to 1.8.6-p287 [Isaac Kearse] 
* upgrade to rubygems 1.2.0 - addresses memory utilization issues on
  low-memory boxes (half-gig or less) [thewoolleyman]
* deprec:rails:install_stack now installs mysql  [mbailey]
* turned tcp_nopush back on [mbailey]
* Correctly handle tarballs with .tgz extension [mbailey]

lib/deprec/recipes/rails.rb
# Copyright 2006-2008 by Mike Bailey. All rights reserved.
Capistrano::Configuration.instance(:must_exist).load do 

  set :app_user_prefix,  'app_'
  set(:app_user) { app_user_prefix + application }
  set :app_group_prefix,  'app_'
  set(:app_group) { app_group_prefix + application }
  set(:app_user_homedir) { deploy_to }
  set :database_yml_in_scm, true
  set :app_symlinks, nil
  set :rails_env, 'production'
  set :gems_for_project, nil # Array of gems to be installed for app
  set :packages_for_project, nil # Array of packages to be installed for app
  set :shared_dirs, nil # Array of directories that should be created under shared/
                        # and linked to in the project

  # Hook into the default capistrano deploy tasks
  before 'deploy:setup', :except => { :no_release => true } do
    top.deprec.rails.setup_user_perms
    top.deprec.rails.create_app_user_and_group
    top.deprec.rails.setup_paths
    top.deprec.rails.setup_shared_dirs
    top.deprec.rails.install_packages_for_project
    top.deprec.rails.install_gems_for_project
    sudo "chown -R #{user}:#{group} #{deploy_to}"
  end

  after 'deploy:setup', :except => { :no_release => true } do
    puts "create_config_dir"
    top.deprec.rails.create_config_dir
    puts "config_gen"
    top.deprec.rails.config_gen
    top.deprec.rails.config
    top.deprec.rails.activate_services
    top.deprec.rails.set_perms_on_shared_and_releases
    top.deprec.web.reload
    top.deprec.rails.setup_database
  end

  after 'deploy:symlink', :roles => :app do
    top.deprec.rails.symlink_shared_dirs
    top.deprec.rails.symlink_database_yml unless database_yml_in_scm
    top.deprec.rails.make_writable_by_app
    top.deprec.passenger.set_owner_of_environment_rb if app_server_type.to_s == 'passenger' or app_server_type.to_s == 'mod_rails'
  end

  after :deploy, "deploy:cleanup"
  
  namespace :rake do
    namespace :gems do
      task :install, :roles => :app do
        run "cd #{current_path} && #{sudo} rake gems:install RAILS_ENV=#{stage}"
      end
    end
  end
  namespace :db do
    task :migrate, :roles => :db do
      run "cd #{current_path} && #{sudo} rake gems:install RAILS_ENV=#{stage}"
    end
  end
  
  namespace :deprec do
    namespace :rails do
      
      #
      # If database.yml is not kept in scm and it is present in local
      # config dir then push it out to server.
      #
      before 'deprec:rails:symlink_database_yml', :roles => :app do
        push_database_yml unless database_yml_in_scm
      end
      
      task :setup_database, :roles => :db do
        if ! roles[:db].servers.empty? # Some apps don't use database!
          deprec2.read_database_yml
          top.deprec.db.create_user
          top.deprec.db.create_database
          top.deprec.db.grant_user_access_to_database
        end
      end
      
      task :install, :roles => :app do
        install_deps
        install_gems
      end

      task :install_deps do
        apps = case db_server_type
          when :mysql then 'libmysqlclient15-dev'
          when :postgresql then 'libpq-dev'
          when :sqlite then 'sqlite3 libsqlite3-ruby libsqlite3-dev'
          else 'libmysqlclient15-dev sqlite3 libsqlite3-ruby libsqlite3-dev libpq-dev'
        end
        apt.install( {:base => apps}, :stable )
      end
      
      # install some required ruby gems
      task :install_gems do
        gem2.install 'sqlite3-ruby' if db_server_type == :sqlite
        gem2.install 'mysql' if db_server_type == :mysql
        gem2.install 'ruby-pg' if db_server_type == :postgresql
        gem2.install 'rails' unless app_server_type == :mod_rails
        gem2.install 'rake' unless app_server_type == :mod_rails
        gem2.install 'rspec'
      end
      
      desc <<-DESC
      Install full rails stack on a stock standard ubuntu server (7.10, 8.04)
      DESC
      task :install_stack do   

        top.deprec.ruby.install
        top.deprec.rails.install
        top.deprec.svn.install
        top.deprec.git.install
        top.deprec.web.install        # Uses value of web_server_type 
        top.deprec.app.install        # Uses value of app_server_type
        top.deprec.monit.install
        top.deprec.logrotate.install  
        
        # We not longer install database server as part of this task.
        # There is too much danger that someone will wreck an existing
        # shared database.
        #
        # Install database server with:
        #
        #   cap deprec:db:install
      end
      
      task :install_rails_stack do
        puts "deprecated: this task is now called install_stack"
        install_stack
      end
      
      desc "Generate config files for rails app."
      task :config_gen do
        top.deprec.web.config_gen_project unless app_server_type == :mod_rails
        top.deprec.app.config_gen_project
      end

      desc "Push out config files for rails app."
      task :config do
        top.deprec.web.config_project unless app_server_type == :mod_rails
        top.deprec.app.config_project
      end

      task :create_config_dir, :roles => :app do
        deprec2.mkdir("#{shared_path}/config", :group => group, :mode => 0775, :via => :sudo)
      end
      
      # XXX This should be restricted a bit to limit what app can write to. - Mike
      desc "set group ownership and permissions on dirs app server needs to write to"
      task :make_writable_by_app, :roles => :app do
        dirs = "#{shared_path} #{current_path}/tmp #{current_path}/public"
        sudo "chgrp -R #{app_group} #{dirs}"
        sudo "chmod -R g+w #{dirs}" 
        
        # set owner and group of log files 
        # XXX Factor out mongrel
        files = ["#{mongrel_log_dir}/mongrel.log", "#{mongrel_log_dir}/#{rails_env}.log"]
        files.each { |file|
          sudo "touch #{file}"
          sudo "chown #{app_user} #{file}"   
          sudo "chgrp #{app_group} #{file}" 
          sudo "chmod g+w #{file}"   
        } 
      end
      
      desc "Create deployment group and add current user to it"
      task :setup_user_perms, :roles => [:app, :web] do
        deprec2.groupadd(group)
        deprec2.add_user_to_group(user, group)
        deprec2.groupadd(app_group)
        deprec2.add_user_to_group(user, app_group)
        # we've just added ourself to a group - need to teardown connection
        # so that next command uses new session where we belong in group 
        deprec2.teardown_connections
      end
      
      desc "Create user and group for application to run as"
      task :create_app_user_and_group, :roles => :app do
        deprec2.groupadd(app_group) 
        deprec2.useradd(app_user, :group => app_group, :homedir => false)
        # Set the primary group for the user the application runs as (in case 
        # user already existed when previous command was run)
        sudo "usermod --gid #{app_group} --home #{app_user_homedir} #{app_user}"
      end
      
      task :set_perms_on_shared_and_releases, :roles => :app do
        releases = File.join(deploy_to, 'releases')
        sudo "chgrp -R #{group} #{shared_path} #{releases}"
        sudo "chmod -R g+w #{shared_path} #{releases}"
      end

      # Setup database server.
      task :setup_db, :roles => :db, :only => { :primary => true } do
        top.deprec.mysql.setup
      end

      # setup extra paths required for deployment
      task :setup_paths, :roles => [:app, :web] do
        deprec2.mkdir(deploy_to, :mode => 0775, :group => group, :via => :sudo)
        deprec2.mkdir(shared_path, :mode => 0775, :group => group, :via => :sudo)
      end
      
      # Symlink list of files and dirs from shared to current
      #
      # XXX write up explanation
      #
      desc "Setup shared dirs"
      task :setup_shared_dirs, :roles => [:app, :web] do
        if shared_dirs
          shared_dirs.each { |dir| deprec2.mkdir( "#{shared_path}/#{dir}", :via => :sudo ) }
        end
      end
      #
      desc "Symlink shared dirs."
      task :symlink_shared_dirs, :roles => [:app, :web] do
        if shared_dirs
          shared_dirs.each do |dir| 
            path = File.split(dir)[0]
            if path != '.'
              deprec2.mkdir("#{current_path}/#{path}")
            end
            sudo "test -d #{current_path}/#{dir} && mv #{current_path}/#{dir} #{current_path}/#{dir}.moved_by_deprec; exit 0"
            run "ln -nfs #{shared_path}/#{dir} #{current_path}/#{dir}" 
          end
        end
      end
      
      task :install_packages_for_project, :roles => :app do
        if packages_for_project
          apt.install({ :base => packages_for_project }, :stable)
        end
      end
      
      task :install_gems_for_project, :roles => :app do
        if gems_for_project
          gems_for_project.each { |gem| gem2.install(gem) }
        end
      end
      
      desc "Activate web, app and monit"
      task :activate_services do
        top.deprec.web.activate unless app_server_type == :mod_rails      
        top.deprec.app.activate
        top.deprec.monit.activate
      end

      # database.yml stuff
      #
      # XXX DRY this up 
      # I don't know how to let :gen_db_yml check if values have been set.
      #
      # if (self.respond_to?("db_host_#{rails_env}".to_sym)) # doesn't seem to work

      set :db_host_default, lambda { Capistrano::CLI.prompt 'Enter database host', 'localhost'}
      set :db_host_staging, lambda { db_host_default }
      set :db_host_production, lambda { db_host_default }

      set :db_name_default, lambda { Capistrano::CLI.prompt 'Enter database name', "#{application}_#{rails_env}" }
      set :db_name_staging, lambda { db_name_default }
      set :db_name_production, lambda { db_name_default }

      set :db_user_default, lambda { Capistrano::CLI.prompt 'Enter database user', 'root' }
      set :db_user_staging, lambda { db_user_default }
      set :db_user_production, lambda { db_user_default }

      set :db_pass_default, lambda { Capistrano::CLI.prompt 'Enter database pass', '' }
      set :db_pass_staging, lambda { db_pass_default }
      set :db_pass_production, lambda { db_pass_default }

      set :db_adaptor_default, lambda { Capistrano::CLI.prompt 'Enter database adaptor', 'mysql' }
      set :db_adaptor_staging, lambda { db_adaptor_default }
      set :db_adaptor_production, lambda { db_adaptor_default }

      set :db_socket_default, lambda { Capistrano::CLI.prompt('Enter database socket', '')}
      set :db_socket_staging, lambda { db_socket_default }
      set :db_socket_production, lambda { db_socket_default }

      task :generate_database_yml, :roles => :app do    
        database_configuration = render :template => <<-EOF
        #{rails_env}:
        adapter: #{self.send("db_adaptor_#{rails_env}")}
        database: #{self.send("db_name_#{rails_env}")}
        username: #{self.send("db_user_#{rails_env}")}
        password: #{self.send("db_pass_#{rails_env}")}
        host: #{self.send("db_host_#{rails_env}")}
        socket: #{self.send("db_socket_#{rails_env}")}
        EOF
        run "mkdir -p #{deploy_to}/#{shared_dir}/config" 
        put database_configuration, "#{deploy_to}/#{shared_dir}/config/database.yml" 
      end

      desc "Link in the production database.yml" 
      task :symlink_database_yml, :roles => :app do
        run "#{sudo} ln -nfs #{shared_path}/config/database.yml #{current_path}/config/database.yml" 
      end
      
      desc "Copy database.yml to shared/config/database.yml. Useful if not kept in scm"
      task :push_database_yml, :roles => :app do
        stage = exists?(:stage) ? fetch(:stage).to_s : ''
        full_local_path = File.join('config', stage, 'database.yml')
        if File.exists?(full_local_path)
          std.su_put(File.read(full_local_path), "#{shared_path}/config/database.yml", '/tmp/')
        end
      end
      
    end

    namespace :database do
      
      desc "Create database"
      task :create, :roles => :app do
        run "cd #{deploy_to}/current && rake db:create RAILS_ENV=#{rails_env}"
      end

      desc "Run database migrations"
      task :migrate, :roles => :app do
        run "cd #{deploy_to}/current && rake db:migrate RAILS_ENV=#{rails_env}"
      end
      
      desc "Run database migrations"
      task :schema_load, :roles => :app do
        run "cd #{deploy_to}/current && rake db:schema:load RAILS_ENV=#{rails_env}"
      end

      desc "Roll database back to previous migration"
      task :rollback, :roles => :app do
        run "cd #{deploy_to}/current && rake db:rollback RAILS_ENV=#{rails_env}"
      end

    end

  end
  
end


